Noisy `Could not resolve Gradle distribution sources` during sync

## Summary

When the Gradle distribution sources cannot be resolved (e.g. running a snapshot/nightly or locally built Gradle whose `-src.zip` isn't published), the warning

```
Could not resolve Gradle distribution sources. See debug logs for details.
```

is logged many times during a single IDE sync / tooling-model fetch, and the (expensive) repository resolution is re-attempted each time. The failure is non-fatal — `SourceDistributionResolver.sourceDirs()` catches it and returns an empty list — but the repeated warnings are noisy and the repeated resolution attempts are wasted work.

<img width="4470" height="1874" alt="Image" src="https://github.com/user-attachments/assets/d21466ad-9c5d-4c5c-b03f-36b2c5aee9ae" />


## Proposed fix

Cache the resolution outcome (including failure) at a build-scoped level so it's attempted at most once per build and the warning is logged once. 

Slack thread:
- https://gradle.slack.com/archives/CDH5M7FAT/p1781076344444519?thread_ts=1780578050.753499&cid=CDH5M7FAT

## Steps to reproduce

1. Run a Gradle whose distribution sources can't be resolved — e.g. a snapshot/nightly or locally built Gradle (here `9.7.0-20260609103035+0000`) whose `-src.zip` is not on `services.gradle.org/distributions-snapshots`.
2. Open a multi-project Gradle build in IntelliJ IDEA using that distribution.
3. Trigger a Gradle sync / project import.
4. Observe the `Could not resolve Gradle distribution sources` `WARN` logged repeatedly (once per `.gradle.kts` model build); full stacktrace only at `DEBUG`.

## Environment

- Gradle: `9.7.0-20260609103035+0000` (snapshot)
- IDE: IntelliJ IDEA (`com.intellij.gradle.toolingExtension` model fetch)
- Triggered from: `KotlinDslScriptsModelBuilder` during tooling-model build
- Observed: 2026-06-10

## Stacktrace (DEBUG)

<details>
<summary>Gradle distribution sources resolution failure — logged once per script-model build</summary>

```
org.gradle.api.GradleException: Unable to resolve Gradle distribution sources, tried:
  - https://services.gradle.org/distributions-snapshots
    at org.gradle.kotlin.dsl.resolver.SourceDistributionResolver.resolveSourceDirsFromRepositories(SourceDistributionProvider.kt:90)
    at org.gradle.kotlin.dsl.resolver.SourceDistributionResolver.getSourceDirs(SourceDistributionProvider.kt:59)
    at org.gradle.kotlin.dsl.resolver.SourcePathProvider.downloadedSources(SourcePathProvider.kt:82)
    at org.gradle.kotlin.dsl.resolver.SourcePathProvider.sourcePathFor(SourcePathProvider.kt:43)
    at org.gradle.kotlin.dsl.tooling.builders.KotlinScriptTargetModelBuilder.gradleSource(KotlinBuildScriptModelBuilder.kt:401)
    at org.gradle.kotlin.dsl.tooling.builders.KotlinScriptTargetModelBuilder.buildModel(KotlinBuildScriptModelBuilder.kt:382)
    at org.gradle.kotlin.dsl.tooling.builders.KotlinDslScriptsModelBuilder.buildScriptModel(KotlinDslScriptsModelBuilder.kt:119)
    ... (IntelliJ GradleModelFetchAction -> tooling API -> daemon) ...
    Suppressed: org.gradle.api.internal.artifacts.ivyservice.TypedResolveException: Could not resolve all files for configuration 'detachedConfiguration1'.
        at org.gradle.kotlin.dsl.resolver.SourceDistributionResolver.resolveSourceDirsFrom(SourceDistributionProvider.kt:104)
        ...
    Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find gradle:gradle:9.7.0-20260609103035+0000.
    Searched in the following locations:
      - https://services.gradle.org/distributions-snapshots/gradle-9.7.0-20260609103035+0000-src.zip
    Required by:
        detached context for root project 'junit-framework'
```

</details>

--- a/platforms/core-configuration/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/resolver/SourceDistributionResolverIntegrationTest.kt
+++ b/platforms/core-configuration/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/resolver/SourceDistributionResolverIntegrationTest.kt
@@ -10,6 +10,8 @@ import org.gradle.test.fixtures.server.http.HttpServer
 import org.gradle.test.precondition.Requires
 import org.gradle.test.preconditions.TestExecutionPreconditions
 import org.gradle.util.GradleVersion
+import org.hamcrest.CoreMatchers.equalTo
+import org.hamcrest.MatcherAssert.assertThat
 import org.junit.Assume.assumeTrue
 import org.junit.Test
 import java.io.File
@@ -384,6 +386,47 @@ class SourceDistributionResolverIntegrationTest : AbstractKotlinIntegrationTest(
             } finally {
                 primaryServer.stop()
                 fallbackServer.stop()
+            }
+        }
+    }
+
+    @Test
+    @LeaksFileHandles
+    @Requires(TestExecutionPreconditions.NotEmbeddedExecutor::class, reason = "srcDistribution is only available in forked mode")
+    fun `warning about unresolvable distribution sources is logged only once per build`() {
+        val gradleVersion = distribution.version
+        val repositoryName = gradleVersion.repositoryName
+        val artifactFileName = gradleVersion.artifactFileName
+
+        withOwnGradleUserHomeDir("need fresh cache for artifact resolution") {
+            val primaryServer = HttpServer().apply { start() }
+            val fallbackServer = HttpServer().apply { start() }
+            try {
+                withCustomGradleProperties("${primaryServer.uri}/$repositoryName/gradle-${gradleVersion.version}-bin.zip")
+
+                // Both repositories permanently fail to serve the sources, any number of attempts
+                primaryServer.allowGetOrHeadMissing("/$repositoryName/$artifactFileName")
+                fallbackServer.allowGetOrHeadMissing("/$repositoryName/$artifactFileName")
+
+                // The IDE builds one Kotlin DSL model per script, each with its own resolver instance.
+                // Resolving twice in one build must still warn only once and resolve only once.
+                withBuildScript(
+                    """
+                    ${SourceDistributionResolver::class.qualifiedName}(project).sourceDirs()
+                    ${SourceDistributionResolver::class.qualifiedName}(project).sourceDirs()
+                    """
+                )
+
+                val result = build("-Dorg.gradle.kotlin.dsl.resolver.defaultGradleDistRepoBaseUrl=${fallbackServer.uri}")
+
+                val warning = "Could not resolve Gradle distribution sources. See debug logs for details."
+                val warningCount = result.output.split(warning).size - 1
+                assertThat(warningCount, equalTo(1))
+            } finally {
+                primaryServer.stop()
+                fallbackServer.stop()
+                primaryServer.resetExpectations()
+                fallbackServer.resetExpectations()
             }
         }
     }

--- a/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/SourceDistributionProvider.kt
+++ b/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/SourceDistributionProvider.kt
@@ -28,6 +28,8 @@ import org.gradle.internal.Try
 import org.gradle.kotlin.dsl.create
 import org.gradle.kotlin.dsl.resolver.internal.GradleDistRepoDescriptor
 import org.gradle.kotlin.dsl.resolver.internal.GradleDistRepoDescriptorLocator
+import org.gradle.kotlin.dsl.resolver.internal.SourceDistributionResolutionCache
+import org.gradle.kotlin.dsl.support.serviceOf
 import java.io.File
 import kotlin.jvm.optionals.getOrNull
 
@@ -47,18 +49,16 @@ class SourceDistributionResolver(private val project: Project) : SourceDistribut
     private val repoLocator = GradleDistRepoDescriptorLocator(project)
 
     override fun sourceDirs(): Collection<File> =
-        try {
-            sourceDirs
-        } catch (ex: Exception) {
-            project.logger.warn("Could not resolve Gradle distribution sources. See debug logs for details.")
-            project.logger.debug("Gradle distribution sources resolution failure", ex)
-            emptyList()
+        // Cached for the duration of the build so the resolution, and its failure warning, happen at most once
+        project.serviceOf<SourceDistributionResolutionCache>().computeIfAbsent {
+            try {
+                resolveSourceDirsFromRepositories(candidateRepositories())
+            } catch (ex: Exception) {
+                project.logger.warn("Could not resolve Gradle distribution sources. See debug logs for details.")
+                project.logger.debug("Gradle distribution sources resolution failure", ex)
+                emptyList()
+            }
         }
-
-    private
-    val sourceDirs by lazy {
-        resolveSourceDirsFromRepositories(candidateRepositories())
-    }
 
     private
     fun candidateRepositories() =

--- a/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/internal/SourceDistributionResolutionCache.kt
+++ b/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/internal/SourceDistributionResolutionCache.kt
@@ -1,0 +1,40 @@
+/*
+ * Copyright 2026 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.resolver.internal
+
+import org.gradle.internal.service.scopes.Scope
+import org.gradle.internal.service.scopes.ServiceScope
+import java.io.File
+
+
+/**
+ * Memoizes Gradle distribution source resolution, which is invariant across the build tree.
+ *
+ * The IDE creates a fresh resolver per script model, so without this the network-bound resolution, and its
+ * failure warning, would repeat for every script. Build-tree scoped, so a later build retries.
+ */
+@ServiceScope(Scope.BuildTree::class)
+internal class SourceDistributionResolutionCache {
+
+    @Volatile
+    private var cachedSourceDirs: Collection<File>? = null
+
+    fun computeIfAbsent(resolve: () -> Collection<File>): Collection<File> =
+        cachedSourceDirs ?: synchronized(this) {
+            cachedSourceDirs ?: resolve().also { cachedSourceDirs = it }
+        }
+}

--- a/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/services/KotlinScriptServices.kt
+++ b/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/services/KotlinScriptServices.kt
@@ -32,6 +32,7 @@ class KotlinScriptServices : AbstractGradleModuleServices() {
     override fun registerBuildTreeServices(registration: ServiceRegistration) {
         registration.addProvider(org.gradle.kotlin.dsl.accessors.BuildTreeServices)
         registration.addProvider(org.gradle.kotlin.dsl.support.BuildTreeServices)
+        registration.add(org.gradle.kotlin.dsl.resolver.internal.SourceDistributionResolutionCache::class.java)
     }
 
     override fun registerGlobalServices(registration: ServiceRegistration) {

